src/lua/datetime.lua

```diff
@@ -595,102 +603,34 @@ local function datetime_tostring(self)
-    else
-        nsec = NANOS_PER_SEC - nsec % NANOS_PER_SEC
-    end
-    return sec, nsec
-end
-
--- signed or unsigned textual representation of sec and nsec
-local function seconds_str(need_sign, sec, nsec)
-    sec = math_fmod(sec, 60)
-    sec, nsec = denormalize_interval_nsec(sec, nsec)
-    local is_negative = sec < 0 or (sec == 0 and nsec < 0)
-    local str = is_negative and '-' or (need_sign and '+' or '')
-    sec = math_abs(sec)
-    if nsec ~= 0 then
-        str = str .. ('%d.%09d seconds'):format(sec, math_abs(nsec))
-    else
-        str = str .. ('%d seconds'):format(sec)
-    end
-    return str
-end
-
-
-local signed_fmt = {
-    [false] = '%d',
-    [true]  = '%+d',
-}
-
 --[[
     Convert to text interval values of different types
 
     - depending on a values stored there generic interval
-      values may display in following format:
+      values may be displayed in the following format:
         +12 secs
         -23 minutes, 0 seconds
         +12 hours, 23 minutes, 1 seconds
         -7 days, -23 hours, -23 minutes, -1 seconds
     - years will be displayed as
         +10 years
     - months will be displayed as:
-         +2 months
+        +2 months
 ]]
-local function interval_tostring(o)
-    check_interval(o, 'datetime.interval.tostring')
-    local s = ''
-    local need_sign = true
-    if o.year ~= 0 then
-        s = qtail(s) .. ('%+d years'):format(o.year)
-        need_sign = false
-    end
-    if o.month ~= 0 then
-        s = qtail(s) .. (signed_fmt[need_sign] .. ' months'):format(o.month)
-        need_sign = false
-    end
-    local sec, nsec = o.sec, o.nsec
-    if sec == 0 and nsec == 0 then
-        return #s > 0 and s or '0 seconds'
-    end
-    local abs_s = math_abs(sec)
-    if abs_s < 60 then
-        s = qtail(s) .. seconds_str(need_sign, sec, nsec)
-    elseif abs_s < (60 * 60) then
-        s = qtail(s) .. (signed_fmt[need_sign] .. ' minutes, %s'):
-                        format(o.min, seconds_str(false, sec, nsec))
-    elseif abs_s < SECS_PER_DAY then
-        s = qtail(s) .. (signed_fmt[need_sign] .. ' hours, %d minutes, %s'):
-                        format(o.hour, math_fmod(o.min, 60),
-                               seconds_str(false, sec, nsec))
-    else
-        s = qtail(s) .. (signed_fmt[need_sign] .. ' days, %d hours, %d minutes, %s'):
-                        format(o.day, math_fmod(o.hour, 24),
-                               math_fmod(o.min, 60),
-                               seconds_str(false, sec, nsec))
+local function interval_tostring(self)
+    check_interval(self, 'datetime.interval.tostring')
+    local buff = ival_tostr_stash_take()
+    local len = builtin.tnt_interval_to_string(self, buff, IVAL_TOSTRING_BUFSIZE)
+    if len < IVAL_TOSTRING_BUFSIZE then
+        local s = ffi.string(buff)
+        ival_tostr_stash_put(buff)
+        return s
     end
-    return s
+    -- slow path - reallocate for a fuller size, and then restart interval_to_string
```

Comment:
Because such huge buffer is unnecessary in Lua in 99% of a cases. So there is no point to increase size of a pool memory 2.5 times without any need. (And there is graceful scenario implemented to handle rare cases when it was not enough for resultant value)

Generally, it's ok to have different buffer sizes in Lua and C code. IMVHO.